### PR TITLE
[factorio-mod-portal] services

### PR DESCRIPTION
**Pull request by kuanoni**
Saw this https://github.com/badges/shields/issues/8582 issue and thought I'd take a crack at closing it. I'm new to open-source contributions so I'm eager to hear any feedback/changes.

image

API: `https://mods.factorio.com/api/mods/{name}` where {name} is the name of the mod, e.g.
https://mods.factorio.com/api/mods/rso-mod

No authentication is required.
API is documented here: https://wiki.factorio.com/Mod_portal_API#/api/mods/

### CODE REVIEW

File: services/factorio-mod-portal/factorio-mod-portal.service.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
+import Joi from 'joi'
+import { BaseJsonService } from '../index.js'
+import { age } from '../color-formatters.js'
+import { formatDate } from '../text-formatters.js'
+import { renderDownloadsBadge } from '../downloads.js'
+import { renderVersionBadge } from '../version.js'
+
+const schema = Joi.object({
+  downloads_count: Joi.number().required(),
```
Comment (chris48s) [inline services/factorio-mod-portal/factorio-mod-portal.service.js:9]: Lets validate this using `nonNegativeInteger`

File: services/factorio-mod-portal/factorio-mod-portal.service.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    .min(1)
+    .required(),
+}).required()
+
+const queryParamSchema = Joi.object({
+  range: Joi.equal(''),
+}).required()
+
+// Factorio Mod portal API
+// @see https://wiki.factorio.com/Mod_portal_API
```
Comment (chris48s) [inline services/factorio-mod-portal/factorio-mod-portal.service.js:29]: :+1:

File: services/factorio-mod-portal/factorio-mod-portal.service.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  async handle({ modName }) {
+    const { latest_release } = await this.fetch({ modName })
+    return this.constructor.render({ version: latest_release.version })
+  }
+}
+
+// Badge for mod's latest compatible Factorio version
+// Query 'range' to display range of compatible versions
+class FactorioModPortalFactorioVersions extends BaseFactorioModPortalService {
+  static category = 'version'
```
Comment (chris48s) [inline services/factorio-mod-portal/factorio-mod-portal.service.js:80]: Category for this one should be `static category = 'platform-support'`

File: services/factorio-mod-portal/factorio-mod-portal.service.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    }
+  }
+}
+
+// Badge for mod's latest updated version
+class FactorioModPortalLatestVersion extends BaseFactorioModPortalService {
+  static category = 'version'
+
+  static route = {
+    base: 'factorio-mod-portal/latest-version',
```
Comment (chris48s) [inline services/factorio-mod-portal/factorio-mod-portal.service.js:53]: Normally we use `/v` as the route for version badges - see https://github.com/badges/shields/blob/master/doc/badge-urls.md

File: services/factorio-mod-portal/factorio-mod-portal.service.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    return this.constructor.render({ last_updated: latest_release.released_at })
+  }
+}
+
+// Badge for mod's total download count
+class FactorioModPortalDownloads extends BaseFactorioModPortalService {
+  static category = 'downloads'
+
+  static route = {
+    base: 'factorio-mod-portal/downloads',
```
Comment (chris48s) [inline services/factorio-mod-portal/factorio-mod-portal.service.js:167]: Normally we use `/dt` as the route for a badge showing total downloads - see https://github.com/badges/shields/blob/master/doc/badge-urls.md

File: services/factorio-mod-portal/factorio-mod-portal.tester.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import {
+  isComposerVersion,
```
Comment (chris48s) [inline services/factorio-mod-portal/factorio-mod-portal.tester.js:2]: `isComposerVersion` is a fairly specific validator for PHP package ranges. It looks like what you really want here is probably `isVPlusDottedVersionNClauses` or `isVPlusDottedVersionAtLeastOne`.

File: services/factorio-mod-portal/factorio-mod-portal.service.js [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+      versions = earliest
+    } else {
+      versions = `${earliest}-${latest}`
+    }
+    return versions
+  }
+
+  async handle({ modName }, queryParams) {
+    const { first_release, latest_release } = await this.fetch({ modName })
+    const range = queryParams.range !== undefined
```
Comment (chris48s) [inline services/factorio-mod-portal/factorio-mod-portal.service.js:117]: Generally for platform & version support badges, we just show compatibility with the base platform for the latest release of the target package/plugin/whatever (in this case, the factorio version(s) that the latest release of the mod is compatible with). I think I'd be in favour of dropping the `range` param and sticking with that here.

File: services/factorio-mod-portal/factorio-mod-portal.service.js [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+      url: `https://mods.factorio.com/api/mods/${modName}`,
+      errorMessages: {
+        404: 'mod not found',
+      },
+    })
+
+    return {
+      downloads_count,
+      first_release: releases[0],
+      latest_release: releases[releases.length - 1],
```
Comment (chris48s) [inline services/factorio-mod-portal/factorio-mod-portal.service.js:43]: Do we have complete confidence this array is sorted by version? The docs aren't completely clear.

Comment (kuanoni) [reply services/factorio-mod-portal/factorio-mod-portal.service.js:39]: I couldn't find in the docs or anywhere else that explicitly says it's
 sorted by version, but I made a small script to run through every mod listed and can confirm that the `releases` array is always sorted by version.

**Review by chris48s — APPROVED** [review]
Nice. This one's good to go. Great first contribution

**Pull request merged**